crates/bevy_ecs/src/change_detection.rs

```diff
@@ -0,0 +1,190 @@
+            ///
+            /// **Note**: This operation is irreversible.
+            #[inline]
+            fn set_changed(&mut self) {
+                self.ticks.set_changed();
+            }
+
+            /// Get the underlying value.
+            /// Does not mark `self` as changed since access is immutable.
+            #[inline]
+            fn get(&self) -> &Self::Target {
+                self.value
+            }
+
+            /// Get the underlying value and mark `self` as "changed".
+            #[inline]
+            fn get_mut(&mut self) -> &mut Self::Target {
+                self.set_changed();
+                self.value
+            }
+
+            /// Get the underlying value **without** marking `self` as "changed".
+            #[inline]
+            fn get_mut_untracked(&mut self) -> &mut Self::Target {
+                self.value
+            }
+        }
+
+        impl<$($generics),*: $($traits),*> Deref for $name<$($generics),*>
+        {
+            type Target = $target;
+
+            #[inline]
+            fn deref(&self) -> &Self::Target {
+                self.value
+            }
+        }
+
+        impl<$($generics),*: $($traits),*> DerefMut for $name<$($generics),*>
+        {
+            #[inline]
+            fn deref_mut(&mut self) -> &mut Self::Target {
+                self.get_mut()
+            }
+        }
+
+        impl<$($generics),*: $($traits),*> AsRef<$target> for $name<$($generics),*>
+        {
+            #[inline]
+            fn as_ref(&self) -> &$target {
+                self.deref()
+            }
+        }
+
+        impl<$($generics),*: $($traits),*> AsMut<$target> for $name<$($generics),*>
+        {
+            #[inline]
+            fn as_mut(&mut self) -> &mut $target {
+                self.deref_mut()
+            }
+        }
+    };
+}
+
+macro_rules! impl_into_inner {
+    ($name:ident < $( $generics:tt ),+ >, $target:ty, $($traits:ident),*) => {
+        impl<$($generics),*: $($traits),*> $name<$($generics),*>
+        {
+            /// Consume `self` and return the contained value while marking `self` as "changed".
+            #[inline]
+            pub fn into_inner(mut self) -> &'a mut $target {
+                self.set_changed();
+                self.value
+            }
+        }
+    };
+}
+
+macro_rules! impl_debug {
+    ($name:ident < $( $generics:tt ),+ >, $($traits:ident),*) => {
+        impl<$($generics),*: $($traits),*> std::fmt::Debug for $name<$($generics),*>
+            where T: std::fmt::Debug
+        {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                f.debug_tuple(stringify!($name))
+                    .field(self.value)
+                    .finish()
+            }
+        }
+
+    };
+}
+
+pub(crate) struct Ticks<'a> {
+    pub(crate) component_ticks: &'a mut ComponentTicks,
+    pub(crate) last_change_tick: u32,
+    pub(crate) change_tick: u32,
+}
+
+impl<'a> Ticks<'a> {
```

Comment:
This helper struct increases the load on the optimizer, likely increasing compilation time. Can you inline the implementation of all methods into `change_detection_impl!()`?

Reply:
I would argue that inlining each of the calls and removing the `Ticks<'a>` struct would increase the amount of duplicated code and therefore should be avoided.
Sure its another level of abstraction, and as any level of abstraction would increase compile times, but it overall improves readability and re-usability. 

However, if you do feel strongly about this, I'm happy to inline the calls.

Reply:
Each method of `Ticks` is only referenced a single time inside `change_detection_impl!()`. As such I don't think it improves readability at all.

Reply:
Oh, I think I seem to understand what you're saying, and yes I do agree now :smile:

Sorry so to clarify. 
1)
Are you suggesting removing the `Ticks` struct and inlining the variables (and implementations).
e.g.
```rust
struct ResMut<'a, T> { 
    value: &'a mut T,
    pub(crate) component_ticks: &'a mut ComponentTicks,
    pub(crate) last_change_tick: u32,
    pub(crate) change_tick: u32,
}
```
2)
Or just removing the `impl Ticks` and inlining the `is_added` `is_chaged` etc...

Reply:
I was thinking of the first option. The second option would also work fine though if you prefer that.